crypto/ocsp/ocsp_verify.c

```diff
@@ -0,0 +1,382 @@
+  X509_STORE_CTX *ctx = X509_STORE_CTX_new();
+  int ret = -1;
+
+  if (ctx == NULL) {
+    OPENSSL_PUT_ERROR(OCSP, ERR_R_MALLOC_FAILURE);
+    goto end;
+  }
+  if (!X509_STORE_CTX_init(ctx, st, signer, untrusted)) {
+    OPENSSL_PUT_ERROR(OCSP, ERR_R_X509_LIB);
+    goto end;
+  }
+  if(!X509_STORE_CTX_set_purpose(ctx, X509_PURPOSE_OCSP_HELPER)) {
+    OPENSSL_PUT_ERROR(OCSP, ERR_R_X509_LIB);
+    goto end;
+  }
+
+  // Verify |X509_STORE_CTX| and return certificate chain
+  ret = X509_verify_cert(ctx);
+  if (ret <= 0) {
+    ret = X509_STORE_CTX_get_error(ctx);
+    OPENSSL_PUT_ERROR(OCSP, OCSP_R_CERTIFICATE_VERIFY_ERROR);
+    goto end;
+  }
+  if (chain != NULL) {
+    *chain = X509_STORE_CTX_get1_chain(ctx);
+  }
+
+end:
+  X509_STORE_CTX_free(ctx);
+  return ret;
+}
+
+// Check the issuer certificate IDs for equality. If the issuer IDs all match
+// then we just need to check equality against one of them. If there is a mismatch
+// with the same algorithm then there's no point trying to match any certificates
+// against the issuer, and |*ret| will be set to NULL.
+static int ocsp_check_ids(STACK_OF(OCSP_SINGLERESP) *sresp, OCSP_CERTID **ret) {
+  if (sresp == NULL || ret == NULL) {
+    OPENSSL_PUT_ERROR(OCSP, ERR_R_PASSED_NULL_PARAMETER);
+    return -1;
+  }
+
+  OCSP_CERTID *tmpid, *cid;
+  size_t idcount = sk_OCSP_SINGLERESP_num(sresp);
+  if (idcount <= 0) {
+    OPENSSL_PUT_ERROR(OCSP, OCSP_R_RESPONSE_CONTAINS_NO_REVOCATION_DATA);
+    return -1;
+  }
+  cid = sk_OCSP_SINGLERESP_value(sresp, 0)->certId;
+
+  *ret = NULL;
+  for (size_t i = 1; i < idcount; i++) {
+    tmpid = sk_OCSP_SINGLERESP_value(sresp, i)->certId;
+    // Check to see if IDs match
+    if (OCSP_id_issuer_cmp(cid, tmpid) != 0) {
+      // If algorithm mismatch, let caller deal with it instead
+      if (OBJ_cmp(tmpid->hashAlgorithm->algorithm, cid->hashAlgorithm->algorithm) != 0) {
+        return 2;
+      }
+      return 0;
+    }
+  }
+
+  // All IDs match: only need to check one ID
+  *ret = cid;
+  return 1;
+}
+
+static int ocsp_match_issuerid(X509 *cert, OCSP_CERTID *cid,
+                               STACK_OF(OCSP_SINGLERESP) *sresp) {
+  if (cert == NULL) {
+    OPENSSL_PUT_ERROR(OCSP, ERR_R_PASSED_NULL_PARAMETER);
+    return -1;
+  }
+
+  // If only one ID to match then do it.
+  if (cid) {
+    const EVP_MD *dgst;
+    X509_NAME *iname;
+    unsigned char md[EVP_MAX_MD_SIZE];
+    // Set up message digest for comparison.
+    dgst = EVP_get_digestbyobj(cid->hashAlgorithm->algorithm);
+    if (dgst == NULL) {
+      OPENSSL_PUT_ERROR(OCSP, OCSP_R_UNKNOWN_MESSAGE_DIGEST);
+      return -1;
+    }
+    size_t mdlen = EVP_MD_size(dgst);
+    iname = X509_get_subject_name(cert);
+    if (!X509_NAME_digest(iname, dgst, md, NULL)) {
+      return -1;
+    }
+
+    // Compare message digest with |OCSP_CERTID|
+    if ((cid->issuerNameHash->length != (int)mdlen) || (cid->issuerKeyHash->length != (int)mdlen)) {
+      return 0;
+    }
```

Comment:
Casting `size_t` to `int` seems not safe impl. Just check if `length >= 0` and then do the comparison?